api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -405,11 +428,11 @@ public Builder month(String sourceName) {
     }
 
     public Builder day(String sourceName, String targetName) {
-      checkAndAddPartitionName(targetName);
       Types.NestedField sourceColumn = findSourceColumn(sourceName);
       PartitionField field = new PartitionField(
           sourceColumn.fieldId(), nextFieldId(), targetName, Transforms.day(sourceColumn.type()));
       checkForRedundantPartitions(field);
+      checkAndAddPartitionName(targetName);
```

Comment:
I think this was needed because of my comment that the duplicate field should throw an exception before the duplicate name.